src/descriptor/mod.rs

```diff
@@ -848,7 +978,153 @@ impl<Pk: MiniscriptKey> fmt::Display for Descriptor<Pk> {
             Descriptor::Wsh(ref sub) => write!(f, "wsh({})", sub),
             Descriptor::ShWsh(ref sub) => write!(f, "sh(wsh({}))", sub),
+            Descriptor::ShSortedMulti(ref smv) => write!(f, "sh({})", smv),
+            Descriptor::WshSortedMulti(ref smv) => write!(f, "wsh({})", smv),
+            Descriptor::ShWshSortedMulti(ref smv) => write!(f, "sh(wsh({}))", smv),
+        }
+    }
+}
+
+/// Contents of a "sortedmulti" descriptor
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct SortedMultiVec<Pk: MiniscriptKey, Ctx: ScriptContext> {
+    /// signatures required
+    pub k: usize,
+    /// public keys inside sorted Multi
+    pub pks: Vec<Pk>,
+    /// The current ScriptContext for sortedmulti
+    pub(crate) phantom: PhantomData<Ctx>,
+}
+
+impl<Pk: MiniscriptKey, Ctx: ScriptContext> SortedMultiVec<Pk, Ctx> {
+    fn new(k: usize, pks: &[Pk]) -> Result<Self, Error> {
+        // Check the limits before creating a new SortedMultiVec
+        // For example, under p2sh context the scriptlen can only be
+        // upto 520 bytes.
+        let term: miniscript::decode::Terminal<Pk, Ctx> = Terminal::Multi(k, Vec::from(pks));
+        let ms = Miniscript::from_ast(term)?;
+
+        // This would check all the consensus rules for p2sh/p2wsh and
+        // even tapscript in future
+        Ctx::check_ms_validity(&ms)?;
+
+        Ok(Self {
+            k,
+            pks: Vec::from(pks),
+            phantom: PhantomData,
+        })
+    }
+    /// Parse an expression tree into a SortedMultiVec
+    fn from_tree(tree: &Tree) -> Result<Self, Error>
+    where
+        <Pk as FromStr>::Err: ToString,
+    {
+        let k = tree.args[0].name.parse::<usize>().unwrap(); // FIXME: unwrap
+        let pks: Result<Vec<Pk>, _> = tree.args[1..]
+            .iter()
+            .map(|sub| expression::terminal(sub, Pk::from_str))
+            .collect();
+        pks.map(|pks| SortedMultiVec::new(k, &pks))?
+    }
+}
+
+impl<Pk: MiniscriptKey, Ctx: ScriptContext> SortedMultiVec<Pk, Ctx> {
+    fn translate_pk<FPk, Q, FuncError>(
+        &self,
+        translatefpk: &mut FPk,
+    ) -> Result<SortedMultiVec<Q, Ctx>, FuncError>
+    where
+        FPk: FnMut(&Pk) -> Result<Q, FuncError>,
+        Q: MiniscriptKey,
+    {
+        let pks: Result<Vec<Q>, _> = self.pks.iter().map(&mut *translatefpk).collect();
+        Ok(SortedMultiVec {
+            k: self.k,
+            pks: pks?,
+            phantom: PhantomData,
+        })
+    }
+}
+
+impl<Pk: MiniscriptKey + ToPublicKey, Ctx: ScriptContext> SortedMultiVec<Pk, Ctx> {
+    /// Create Terminal::Multi containing sorted pubkeys
+    pub fn sorted_node(&self) -> Terminal<bitcoin::PublicKey, Ctx> {
+        let mut pks: Vec<bitcoin::PublicKey> =
+            self.pks.iter().map(|pk| pk.to_public_key()).collect();
+        // Sort pubkeys lexigraphically according to BIP 67
+        pks.sort_by(|a, b| a.to_string().partial_cmp(&b.to_string()).unwrap());
```

Comment:
Then with the above, this should just become `pks.sort_by_key(Bip67Sorted)`.

Reply:
Did you mean `pks.sort_by_key(|pk| Bip67Sorted(pk))`?

According to [the docs](https://doc.rust-lang.org/beta/std/vec/struct.Vec.html#method.sort_by_key), the argument to `sort_by_key` needs to be `FnMut`, which we haven't implemented.

If I use a lambda, then it complains that return value of `Bip67Sorted(pk)` may outlive `pk` ... not sure how to fix that.

Reply:
I think it would be best to just use this for now:
```
pks.sort_by(|a, b| {
    a.to_public_key()
        .serialize()
        .partial_cmp(&b.to_public_key().serialize())
        .unwrap()
});
```

But get [the upstream change](https://github.com/rust-bitcoin/rust-secp256k1/pull/248) merged and switch to `pks.sort()`  as soon as upstream change is released.

Reply:
Doesn't make sense to make a new public struct to implement behavior that should be in `rust-libsecp256k1`

Reply:
`pks.sort_by_key(|pk| Bip67Sorted(pk))` is identical to `pks.sort_by_key(Bip67Sorted)` except the latter uses fewer symbols. And struct constructors always implement all of the `Fn` traits.

Using a lambda or not should have no effect on this. What is the error you are getting?

Reply:
I can reproduce the lifetime error. The issue is that `sort_by_key` and `sort_key_cached_key` have their lifetimes set incorrectly by stdlib, so we do indeed need to use `sort_by`

Reply:
Ahh sorry, should have posted the full error message.

Reply:
Unresolving this for now since there is opposition from upstream to including a nonsense Core ordering in rust-bitcoin.